fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/DataInKeySpacePath.java

```diff
@@ -37,24 +38,45 @@
 public class DataInKeySpacePath {
 
     @Nonnull
-    private final CompletableFuture<ResolvedKeySpacePath> resolvedPath;
+    private final KeySpacePath path;
+    @Nullable
+    private final Tuple remainder;
     @Nonnull
     private final byte[] value;
 
-    public DataInKeySpacePath(KeySpacePath path, KeyValue rawKeyValue, FDBRecordContext context) {
-        this.resolvedPath = path.toResolvedPathAsync(context, rawKeyValue.getKey());
-        this.value = rawKeyValue.getValue();
-        if (this.value == null) {
+    public DataInKeySpacePath(@Nonnull final KeySpacePath path, @Nullable final Tuple remainder,
+                              @Nullable final byte[] value) {
+        this.path = path;
+        this.remainder = remainder;
+        if (value == null) {
             throw new RecordCoreArgumentException("Value cannot be null")
-                    .addLogInfo(LogMessageKeys.KEY, ByteArrayUtil2.loggable(rawKeyValue.getKey()));
+                    .addLogInfo(LogMessageKeys.KEY, path);
         }
-    }
-
-    public CompletableFuture<ResolvedKeySpacePath> getResolvedPath() {
-        return resolvedPath;
+        this.value = value;
     }
 
     public byte[] getValue() {
         return this.value;
     }
+
+    @Nonnull
+    public KeySpacePath getPath() {
+        return path;
+    }
+
+    @Nullable
+    public Tuple getRemainder() {
+        return remainder;
+    }
+
+    /**
+     * Converts this data item to a {@link ResolvedKeySpacePath} by resolving the path and applying the remainder.
+     * @param context the context to use for resolving the path
+     * @return a future that completes with the resolved path including the remainder
+     */
+    @Nonnull
+    public CompletableFuture<ResolvedKeySpacePath> getResolvedPath(@Nonnull FDBRecordContext context) {
+        return path.toResolvedPathAsync(context)
```

Comment:
`path.toResolvedPathAsync` can be removed from the interface (only called internally)?